src/openms/include/OpenMS/KERNEL/MSExperiment.h

```diff
@@ -326,11 +353,41 @@ namespace OpenMS
     /// Returns a non-mutable area iterator for all peaks in @p range. If a dimension is empty(), it is ignored (i.e. does not restrict the area)
     ConstAreaIterator areaBeginConst(const RangeManagerType& range, UInt ms_level = 1) const;
 
-    /// Returns an non-mutable invalid area iterator marking the end of an area
+    /// Returns a non-mutable invalid area iterator marking the end of an area
     ConstAreaIterator areaEndConst() const;
 
+
+    std::vector<std::pair<size_t,size_t>> getRangesIdcs_(const std::vector<std::pair<RangeMZ, RangeRT>>& mz_rt_ranges) const;
+    std::pair<Size, Size> getSpectraIdxRangeByRetentionTime(double start, double end) const;
+    std::vector<Size> getSpectraIdcsByRetentionTime(double start, double end, unsigned int ms_level) const;
+    using AggregatorFunc = std::function<CoordinateType(const MSSpectrum& s, size_t start, size_t end)>;
+    using ReduceFunc = std::function<CoordinateType(CoordinateType&, CoordinateType)>;
+
+    /// Aggregate all spectra in the range given by the begin and end area iterators.
+    /// The aggregation is performed by applying the aggregator function to the peaks in the range, per dimension (RT and MZ).
+    /// 
+    MSExperiment::CoordinateType aggregate(ConstAreaIterator begin, ConstAreaIterator end, AggregatorFunc rt, AggregatorFunc mz) const;
+    MSExperiment::CoordinateType aggregate(ConstAreaIterator begin, ConstAreaIterator end, ReduceFunc rt, AggregatorFunc mz) const;
+    MSExperiment::CoordinateType aggregate(ConstAreaIterator begin, ConstAreaIterator end, AggregatorFunc rt, ReduceFunc mz) const;
+    MSExperiment::CoordinateType aggregate(ConstAreaIterator begin, ConstAreaIterator end, ReduceFunc rt, ReduceFunc mz) const;
```

Comment:
I am a bit worried this is too much if we also take e.g., IM into account...

Reply:
but I guess is fine... don't see an easier solution that have those problems

Reply:
You can probably use templates with enable_if or in the future traits to specify what kinds of functions you can use.